staking/staking-specification.md

```diff
@@ -0,0 +1,815 @@
+
+The best way we found to mitigate this danger is with runtime sanity checks. We hardcode the expected slot and offset of each state variable and assert the value every time the logic contract is updated. This is handled in the Staking Contract constructor [here](https://github.com/0xProject/0x-monorepo/blob/3.0/contracts/staking/contracts/src/Staking.sol).
+
+See [LibProxy](https://github.com/0xProject/0x-monorepo/blob/3.0/contracts/staking/contracts/src/libs/LibProxy.sol) for the proxy code.
+
+### 9.2 The Read-Only Proxy
+
+The [read-only proxy](https://github.com/0xProject/0x-monorepo/blob/3.0/contracts/staking/contracts/src/ReadOnlyProxy.sol) is stateless and sits between the Staking Contract Proxy and Staking Contract. It forces every call to be read-only by using a force-revert delegate-call.
+
+Steps:
+1. The read-only proxy delegates the incoming call to itself.
+2. On re-entry, we delegate to the Staking Contract and then _reverts_ with the return data.
+3. The revert is caught and returned to the Staking Proxy.
+
+<p align="center"><img src="./assets/Read-Only%20Proxy.png" width="600" /></p>
+
+See [LibProxy](https://github.com/0xProject/0x-monorepo/blob/3.0/contracts/staking/contracts/src/libs/LibProxy.sol) for the proxy code.
+
+### 9.3 Tracking for Reward Balances for Pool Members
+
+
+This section describes the workflow for tracking and computing the portion of a pool's reward that belongs to a given member. The general equations for this are shown below.
+
+A pool with _D_ delegated stake that earned _R_ rewards for its pool members in a given epoch, the reward (_r_) for a member that delegated _d_ stake is computed by:
+
+<p align="center"><img src="./assets/reward_tracking/RewardSingleEpoch.png" height="60" /></p>
+
+
+The member's reward after n epochs (given member does not change their stake) is then given by:
+
+<p align="center"><img src="./assets/reward_tracking/RewardAfterManyEpochs.png" height="60" /></p>
+
+
+When a member modifies their stake in the pool, the [StoredBalance struct](https://github.com/0xProject/0x-monorepo/blob/3.0/contracts/staking/contracts/src/interfaces/IStructs.sol) gives us:
+1. How many epochs they were staked (`n`)
+2. How much stake they had contributed during those epochs (`d`)
+
+In addition to these values, we also need sum of ratios `R_k / D_k`, for each epoch `k` that the member was delegated. This ratio is available during the pool's finalization of epoch `k`. We are able to do store this information concisely using a cumulative sum of these reward ratios, as follows:
+
+We store the following ratio for each epoch that a reward is earned for the pool:
+<p align="center"><img src="./assets/reward_tracking/WhatWeStore.png" height="60" /></p>
+
+We compute a member's reward using the following equation:
+<p align="center"><img src="./assets/reward_tracking/RewardFromWhatWeStore.png" height="40" /></p>
```

Comment:
I think that it would help with readability to define the function: `d_(x, y)`. I'm reading this as `d_(x, y)` is the stake delegated from epochs `x` to `y`.